feat: Support pausing and unpausing schedules

WillDaSilva opened the pull request:
Closes https://github.com/agronholm/apscheduler/issues/899

peterschutt commented on src/apscheduler/_schedulers/async_.py:
```diff
@@ -30,7 +30,13 @@
 from .._context import current_async_scheduler, current_job
 from .._converters import as_enum, as_timedelta
-from .._enums import CoalescePolicy, ConflictPolicy, JobOutcome, RunState, SchedulerRole
+from .._enums import (
+    CoalescePolicy,
+    ConflictPolicy,
+    JobOutcome,
+    RunState,
+    SchedulerRole,
+)
```
I notice your PR changes the formatting of the imports in a couple of places, seemingly without a need. Just FYI.

WillDaSilva replied:
Addressed in 695efe6

WillDaSilva commented on src/apscheduler/_schedulers/async_.py:
```diff
@@ -30,7 +30,13 @@
 from .._context import current_async_scheduler, current_job
 from .._converters import as_enum, as_timedelta
-from .._enums import CoalescePolicy, ConflictPolicy, JobOutcome, RunState, SchedulerRole
+from .._enums import (
+    CoalescePolicy,
+    ConflictPolicy,
+    JobOutcome,
+    RunState,
+    SchedulerRole,
+)
```
```suggestion
from .._enums import CoalescePolicy, ConflictPolicy, JobOutcome, RunState, SchedulerRole
```

WillDaSilva commented on src/apscheduler/_schedulers/sync.py:
```diff
@@ -6,17 +6,22 @@
 
 from .. import current_scheduler
-from .._enums import CoalescePolicy, ConflictPolicy, RunState, SchedulerRole
+from .._enums import (
+    CoalescePolicy,
+    ConflictPolicy,
+    RunState,
+    SchedulerRole,
+)
```
```suggestion
from .._enums import CoalescePolicy, ConflictPolicy, RunState, SchedulerRole
```

agronholm commented on src/apscheduler/_schedulers/async_.py:
```diff
@@ -529,6 +532,29 @@ async def remove_schedule(self, id: str) -> None:
         self._check_initialized()
         await self.data_store.remove_schedules({id})
 
+    async def pause_schedule(self, id: str) -> None:
```
The `id` parameter shadows the built-in `id()` function. You should rename it to `schedule_id`.

WillDaSilva replied:
@agronholm For this I was just following what was used by other functions, since consistency is valuable for parts of the external API like this. Would you still rather I use `schedule_id` even though the other similar methods use `id`?

WillDaSilva replied:
In 61f7051 I've updated `pause_schedule` and `unpause_schedule` to use `schedule_id` instead of `id`. If you'd rather `id` be used consistently, I can revert that change.

agronholm replied:
You're right, I've been using `id` for other such parameter names. It's okay to revert.

agronholm commented on src/apscheduler/_schedulers/async_.py:
```diff
@@ -529,6 +532,29 @@ async def remove_schedule(self, id: str) -> None:
+    async def pause_schedule(self, id: str) -> None:
+        """Pause the specified schedule."""
+        self._check_initialized()
+        await self.data_store.pause_schedules({id})
```
This code should be in the scheduler (i.e. this method) if it doesn't require any additional functionality from the data store.

WillDaSilva replied:
Good call. Addressed in 61f7051. One consequence of this is that it no longer makes sense to have datastore tests for pausing and unpausing, so those tests have been removed.

agronholm commented on src/apscheduler/_schedulers/async_.py:
```diff
@@ -529,6 +532,29 @@ async def remove_schedule(self, id: str) -> None:
+        until the next fire time is at or after the specified time.
+        """
+        self._check_initialized()
+        await self.data_store.unpause_schedules({id}, resume_from=resume_from)
```
As above, this code should be in the scheduler, not the data store.

WillDaSilva replied:
Addressed in 61f7051

agronholm commented on src/apscheduler/_schedulers/async_.py:
```diff
@@ -529,6 +532,29 @@ async def remove_schedule(self, id: str) -> None:
+        *,
+        resume_from: datetime | Literal["now"] | None = None,
+    ) -> None:
+        """Unpause the specified schedule.
+
+        By default, the schedule will be resumed as if it had never been paused, and all
+        missed runs will be considered misfires. The ``resume_from`` parameter can be
+        used to specify a different time from which to resume the schedule. The string
+        ``'now'`` can be used as shorthand for ``datetime.now(tz=UTC)``. If
+        ``resume_from`` is not ``None``, then the trigger will be repeatedly advanced
+        until the next fire time is at or after the specified time.
```
This is missing Sphinx-style parameter documentations (see other methods for examples). Also, the convention, at least in my projects, is to start the docstring text on a new line after the triple quotes, and add an empty line at the end if the docstring is immediately followed by code.

WillDaSilva replied:
Addressed in 61f7051

agronholm commented on src/apscheduler/datastores/mongodb.py:
```diff
@@ -366,9 +366,19 @@ async def acquire_schedules(self, scheduler_id: str, limit: int) -> list[Schedul
-                            "$or": [
-                                {"acquired_until": {"$exists": False}},
-                                {"acquired_until": {"$lt": now}},
+                            "$and": [
+                                {
+                                    "$or": [
+                                        {"paused": {"$exists": False}},
+                                        {"paused": False},
+                                    ]
+                                },
+                                {
+                                    "$or": [
+                                        {"acquired_until": {"$exists": False}},
+                                        {"acquired_until": {"$lt": now}},
+                                    ]
+                                },
                             ],
```
Do we know if mongo can figure out that it should use the index for `acquired_until`, or do we need to give it some kind of a hint?

agronholm replied:
Oh, never mind, we don't have any index on `acquired_until` either.

agronholm reviewed (changes requested):
Nice job. I'll need some changes though.

WillDaSilva commented on src/apscheduler/datastores/sqlalchemy.py:
```diff
@@ -752,7 +755,10 @@ async def get_next_schedule_run_time(self) -> datetime | None:
-            .where(self._t_schedules.c.next_fire_time.isnot(None))
+            .where(
+                self._t_schedules.c.next_fire_time.isnot(None),
+                self._t_schedules.c.paused.is_(False),
```
Addressed in 61f7051

WillDaSilva commented on src/apscheduler/datastores/sqlalchemy.py:
```diff
@@ -600,6 +602,7 @@ async def acquire_schedules(self, scheduler_id: str, limit: int) -> list[Schedul
                             and_(
                                 self._t_schedules.c.next_fire_time.isnot(None),
                                 comparison,
+                                self._t_schedules.c.paused.is_(False),
```
Addressed in 61f7051

WillDaSilva commented on src/apscheduler/datastores/memory.py:
```diff
@@ -207,8 +207,11 @@ async def acquire_schedules(self, scheduler_id: str, limit: int) -> list[Schedul
-                # The schedule is either paused or not yet due
-                break
+                # The schedule is either exhausted or not yet due
+                continue
```
Ah right, I was mistakenly under the impression that other schedules past this point could be due. I've reverted this to use `break` instead of `continue`, and left a clarifying comment to hopefully help other contributors in the future not make my mistake.

agronholm commented on src/apscheduler/_schedulers/async_.py:
```diff
@@ -529,6 +532,64 @@ async def remove_schedule(self, id: str) -> None:
+            conflict_policy=ConflictPolicy.replace,
+        )
+
+    def _get_unpaused_next_fire_time(
```
This method doesn't seem to be used anywhere but `unpause_schedule()`. Therefore it should be inlined, if there's no justification for it to be a separate method.

agronholm commented on src/apscheduler/_schedulers/async_.py:
```diff
@@ -529,6 +532,64 @@ async def remove_schedule(self, id: str) -> None:
+        schedule: Schedule,
+        resume_from: datetime | Literal["now"] | None,
+    ) -> datetime | None:
+        if resume_from is None:
+            return schedule.next_fire_time
+        if resume_from == "now":
+            resume_from = datetime.now(tz=timezone.utc)
+        if (
+            schedule.next_fire_time is not None
+            and schedule.next_fire_time >= resume_from
+        ):
+            return schedule.next_fire_time
+        try:
+            while (next_fire_time := schedule.trigger.next()) < resume_from:
+                pass  # Advance `next_fire_time` until its at or past `resume_from`
+        except TypeError:  # The trigger is exhausted
+            return None
+        return next_fire_time
```
Add a blank line after each control block to visually distinguish it from others.

agronholm replied:
Also, I don't like catching `TypeError` here, as it has the potential to mask actual bugs. I'd prefer an explicit check against a `None` return value, even if it adds a line or two of code.

agronholm reviewed (changes requested):
Getting there now code-wise. I'll also require a changelog entry (`docs/versionhistory.rst`).

What about the documentation? I think this feature should be documented. Do you think you'd be up for that, or do you want me to do it?

WillDaSilva commented:
@agronholm Thanks for the review. I've addressed your comments related to the code, added an entry in the version history, and added a section about pausing and unpausing schedules to the docs. Let me know if you would like any changes to them, or if you would rather just update them yourself in a follow-up if my writing style in the docs isn't to your liking.

agronholm reviewed (approved):
Thanks, I couldn't have written the docs better myself!
